File: src/main/java/com/klarna/rest/api/model/OrderLine.java
/*
 * Copyright 2014 Klarna AB
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.klarna.rest.api.model;

import java.net.URI;

/**
 * Order line data model.
 */
public class OrderLine extends Model {

    /**
     * Item type.
     */
    private String type;

    /**
     * Article number, SKU or similar.
     */
    private String reference;

    /**
     * Article name.
     */
    private String name;

    /**
     * The item quantity.
     */
    private Long quantity;

    /**
     * Unit used to describe the quantity, e.g. "pcs".
     */
    private String quantityUnit;

    /**
     * Unit price in cents, including tax and excluding discounts.
     */
    private Long unitPrice;

    /**
     * Item tax rate in cents, e.g. 2500 is 25%.
     */
    private Integer taxRate;

    /**
     * Total amount in cents, including tax and discounts.
     */
    private Long totalAmount;

    /**
     * Total discount amount in cents, including tax.
     */
    private Long totalDiscountAmount;

    /**
     * Total tax amount in cents.
     */
    private Long totalTaxAmount;

    /**
     * Data about the order line. Set at creation or update and returned when fetching the order through the API.
     */
    private String merchantData;

    /**
     * Product URI.
     */
    private URI productUrl;

    /**
     * Image URI.
     */
    private URI imageUrl;

    /**
     * Identifiers to better describe the product for improved risk assessment, support,
     * presentation to consumers and promotional functionality.
     */
    private ProductIdentifiers productIdentifiers;

    /**
     * Gets the item type.
     *
     * @return Type
     */
    public String getType() {
        return this.type;
    }

    /**
     * Sets the item type.
     *
     * @param type Type, e.g. "physical", "discount" or "shipping_fee"
     * @return Same instance
     */
    public OrderLine setType(final String type) {
        this.type = type;

        return this;
    }

    /**
     * Gets the item reference.
     *
     * @return Reference
     */
    public String getReference() {
        return this.reference;
    }

    /**
     * Sets the item reference.
     *
     * @param reference Reference
     * @return Same instance
     */
    public OrderLine setReference(final String reference) {
        this.reference = reference;

        return this;
    }

    /**
     * Gets the item name.
     *
     * @return Name
     */
    public String getName() {
        return this.name;
    }

    /**
     * Sets the item name.
     *
     * @param name Name
     * @return Same instance
     */
    public OrderLine setName(final String name) {
        this.name = name;

        return this;
    }

    /**
     * Gets the item quantity.
     *
     * @return Quantity
     */
    public Long getQuantity() {
        return this.quantity;
    }

    /**
     * Sets the item quantity.
     *
     * @param quantity Quantity
     * @return Same instance
     */
    public OrderLine setQuantity(final Long quantity) {
        this.quantity = quantity;

        return this;
    }

    /**
     * Gets the quantity unit.
     *
     * @return Quantity unit, e.g. "pcs"
     */
    public String getQuantityUnit() {
        return this.quantityUnit;
    }

    /**
     * Sets the quantity unit.
     *
     * @param unit Quantity unit, e.g. "pcs"
     * @return Same instance
     */
    public OrderLine setQuantityUnit(final String unit) {
        this.quantityUnit = unit;

        return this;
    }

    /**
     * Gets the unit price.
     *
     * @return Price in cents, e.g. 2500 is 25.00
     */
    public Long getUnitPrice() {
        return this.unitPrice;
    }

    /**
     * Sets the unit price.
     *
     * @param price Price in cents, e.g. 2500 is 25.00
     * @return Same instance
     */
    public OrderLine setUnitPrice(final Long price) {
        this.unitPrice = price;

        return this;
    }

    /**
     * Gets the item tax rate.
     *
     * @return Tax rate in cents, e.g. 2500 is 25%
     */
    public Integer getTaxRate() {
        return this.taxRate;
    }

    /**
     * Sets the item tax rate.
     *
     * @param rate Tax rate in cents, e.g. 2500 is 25%
     * @return Same instance
     */
    public OrderLine setTaxRate(final Integer rate) {
        this.taxRate = rate;

        return this;
    }

    /**
     * Gets total amount.
     *
     * @return Amount in cents, e.g. 2500 is 25.00
     */
    public Long getTotalAmount() {
        return this.totalAmount;
    }

    /**
     * Sets the total amount.
     *
     * @param amount Amount in cents, e.g. 2500 is 25.00
     * @return Same instance
     */
    public OrderLine setTotalAmount(final Long amount) {
        this.totalAmount = amount;

        return this;
    }

    /**
     * Gets the total discount amount.
     *
     * @return Amount in cents, e.g. 2500 is 25.00
     */
    public Long getTotalDiscountAmount() {
        return this.totalDiscountAmount;
    }

    /**
     * Sets the total discount amount.
     *
     * @param amount Amount in cents, e.g. 2500 is 25.00
     * @return Same instance
     */
    public OrderLine setTotalDiscountAmount(final Long amount) {
        this.totalDiscountAmount = amount;

        return this;
    }

    /**
     * Gets the total tax amount.
     *
     * @return Amount in cents, e.g. 2500 is 25.00
     */
    public Long getTotalTaxAmount() {
        return this.totalTaxAmount;
    }

    /**
     * Sets the total tax amount.
     *
     * @param amount Amount in cents, e.g. 2500 is 25.00
     * @return Same instance
     */
    public OrderLine setTotalTaxAmount(final Long amount) {
        this.totalTaxAmount = amount;

        return this;
    }

    /**
     * Gets the merchant metadata.
     *
     * @return Merchant metadata
     */
    public String getMerchantData() {
        return this.merchantData;
    }

    /**
     * Sets the merchant metadata.
     *
     * @param merchantData Merchant metadata
     * @return Same instance
     */
    public OrderLine setMerchantData(final String merchantData) {
        this.merchantData = merchantData;

        return this;
    }

    /**
     * Get the product URI.
     *
     * @return Product URI.
     */
    public URI getProductUrl() {
        return this.productUrl;
    }

    /**
     * Set the product URI.
     *
     * @param productUrl Product URI.
     * @return Same instance.
     */
    public OrderLine setProductUrl(final URI productUrl) {
        this.productUrl = productUrl;

        return this;
    }

    /**
     * Get the image URI.
     *
     * @return Image URI.
     */
    public URI getImageUrl() {
        return this.imageUrl;
    }

    /**
     * Set the image URI.
     *
     * @param imageUrl Image URI.
     * @return Same instance.
     */
    public OrderLine setImageUrl(final URI imageUrl) {
        this.imageUrl = imageUrl;

        return this;
    }

    /**
     * Gets product indentifiers
     *
     * @return productIdentifiers
     **/
    public ProductIdentifiers getProductIdentifiers() {
        return this.productIdentifiers;
    }

    /**
     * Sets product indentifiers
     *
     * @return Same instance.
     **/
    public OrderLine setProductIdentifiers(ProductIdentifiers productIdentifiers) {
        this.productIdentifiers = productIdentifiers;

        return this;
    }
}

File: Klarna.Rest/Klarna.Rest.Core/Model/OrderManagement/OrderLine.cs
using System.Runtime.Serialization;
using System.Text;
using Newtonsoft.Json;

namespace Klarna.Rest.Core.Model.OrderManagement {

  /// <summary>
  /// 
  /// </summary>
  [DataContract]
  public class OrderLine {
    /// <summary>
    /// Article number, SKU or similar.
    /// </summary>
    /// <value>Article number, SKU or similar.</value>
    [DataMember(Name="reference", EmitDefaultValue=false)]
    [JsonProperty(PropertyName = "reference")]
    public string Reference { get; set; }

    /// <summary>
    /// Order line type. Matches: physical|discount|shipping_fee|sales_tax|store_credit|gift_card|digital|surcharge
    /// </summary>
    /// <value>Order line type. Matches: physical|discount|shipping_fee|sales_tax|store_credit|gift_card|digital|surcharge</value>
    [DataMember(Name="type", EmitDefaultValue=false)]
    [JsonProperty(PropertyName = "type")]
    public string Type { get; set; }

    /// <summary>
    /// Item quantity. Non-negative. Between 0 and 100000
    /// </summary>
    /// <value>Item quantity. Non-negative. Between 0 and 100000</value>
    [DataMember(Name="quantity", EmitDefaultValue=false)]
    [JsonProperty(PropertyName = "quantity")]
    public long? Quantity { get; set; }

    /// <summary>
    /// Unit used to describe the quantity. Maximum 10 characters.
    /// </summary>
    /// <value>Unit used to describe the quantity. Maximum 10 characters.</value>
    [DataMember(Name="quantity_unit", EmitDefaultValue=false)]
    [JsonProperty(PropertyName = "quantity_unit")]
    public string QuantityUnit { get; set; }

    /// <summary>
    /// Descriptive item name. Maximum 255 characters.
    /// </summary>
    /// <value>Descriptive item name. Maximum 255 characters.</value>
    [DataMember(Name="name", EmitDefaultValue=false)]
    [JsonProperty(PropertyName = "name")]
    public string Name { get; set; }

    /// <summary>
    /// Total amount including tax and discounts (`quantity * unit_price - total_discount_amount`).
    /// </summary>
    /// <value>Total amount including tax and discounts (`quantity * unit_price - total_discount_amount`).</value>
    [DataMember(Name="total_amount", EmitDefaultValue=false)]
    [JsonProperty(PropertyName = "total_amount")]
    public long? TotalAmount { get; set; }

    /// <summary>
    /// Unit price including tax without applying discounts in minor units.
    /// </summary>
    /// <value>Unit price including tax without applying discounts in minor units.</value>
    [DataMember(Name="unit_price", EmitDefaultValue=false)]
    [JsonProperty(PropertyName = "unit_price")]
    public long? UnitPrice { get; set; }

    /// <summary>
    /// The discount amount in minor units. Includes tax. Example: 1200 = $12. Max value: 100000000
    /// </summary>
    /// <value>The discount amount in minor units. Includes tax. Example: 1200 = $12. Max value: 100000000</value>
    [DataMember(Name="total_discount_amount", EmitDefaultValue=false)]
    [JsonProperty(PropertyName = "total_discount_amount")]
    public long? TotalDiscountAmount { get; set; }

    /// <summary>
    /// The tax rate in percent with two implicit decimals. Non-negative. Example: 2500 = 25%.
    /// </summary>
    /// <value>The tax rate in percent with two implicit decimals. Non-negative. Example: 2500 = 25%.</value>
    [DataMember(Name="tax_rate", EmitDefaultValue=false)]
    [JsonProperty(PropertyName = "tax_rate")]
    public int? TaxRate { get; set; }

    /// <summary>
    /// The total tax amount in minor units. Negative if the order line type is discount. Example: 500 = $5.
    /// </summary>
    /// <value>The total tax amount in minor units. Negative if the order line type is discount. Example: 500 = $5.</value>
    [DataMember(Name="total_tax_amount", EmitDefaultValue=false)]
    [JsonProperty(PropertyName = "total_tax_amount")]
    public long? TotalTaxAmount { get; set; }

    /// <summary>
    /// Data about the order line. Set at creation or update and returned when fetching the order through the API. Maximum 1024 characters.
    /// </summary>
    /// <value>Data about the order line. Set at creation or update and returned when fetching the order through the API. Maximum 1024 characters.</value>
    [DataMember(Name="merchant_data", EmitDefaultValue=false)]
    [JsonProperty(PropertyName = "merchant_data")]
    public string MerchantData { get; set; }

    /// <summary>
    /// URL to the product that can be used in communications between Klarna and the customer. Maximum 1024 characters.
    /// </summary>
    /// <value>URL to the product that can be used in communications between Klarna and the customer. Maximum 1024 characters.</value>
    [DataMember(Name="product_url", EmitDefaultValue=false)]
    [JsonProperty(PropertyName = "product_url")]
    public string ProductUrl { get; set; }

    /// <summary>
    /// URL to an image that can be embedded in communications between Klarna and the customer. Maximum 1024 characters.
    /// </summary>
    /// <value>URL to an image that can be embedded in communications between Klarna and the customer. Maximum 1024 characters.</value>
    [DataMember(Name="image_url", EmitDefaultValue=false)]
    [JsonProperty(PropertyName = "image_url")]
    public string ImageUrl { get; set; }

    /// <summary>
    /// Identifiers to better describe the product for improved risk assessment, support, presentation to consumers and promotional functionality.
    /// </summary>
    /// <value>Identifiers to better describe the product for improved risk assessment, support, presentation to consumers and promotional functionality.</value>
    [DataMember(Name="product_identifiers", EmitDefaultValue=false)]
    [JsonProperty(PropertyName = "product_identifiers")]
    public ProductIdentifiers ProductIdentifiers { get; set; }


    /// <summary>
    /// Get the string presentation of the object
    /// </summary>
    /// <returns>String presentation of the object</returns>
    public override string ToString()  {
      var sb = new StringBuilder();
      sb.Append("class OrderLine {\n");
      sb.Append("  Reference: ").Append(Reference).Append("\n");
      sb.Append("  Type: ").Append(Type).Append("\n");
      sb.Append("  Quantity: ").Append(Quantity).Append("\n");
      sb.Append("  QuantityUnit: ").Append(QuantityUnit).Append("\n");
      sb.Append("  Name: ").Append(Name).Append("\n");
      sb.Append("  TotalAmount: ").Append(TotalAmount).Append("\n");
      sb.Append("  UnitPrice: ").Append(UnitPrice).Append("\n");
      sb.Append("  TotalDiscountAmount: ").Append(TotalDiscountAmount).Append("\n");
      sb.Append("  TaxRate: ").Append(TaxRate).Append("\n");
      sb.Append("  TotalTaxAmount: ").Append(TotalTaxAmount).Append("\n");
      sb.Append("  MerchantData: ").Append(MerchantData).Append("\n");
      sb.Append("  ProductUrl: ").Append(ProductUrl).Append("\n");
      sb.Append("  ImageUrl: ").Append(ImageUrl).Append("\n");
      sb.Append("  ProductIdentifiers: ").Append(ProductIdentifiers).Append("\n");
      sb.Append("}\n");
      return sb.ToString();
    }

    /// <summary>
    /// Get the JSON string presentation of the object
    /// </summary>
    /// <returns>JSON string presentation of the object</returns>
    public string ToJson() {
      return JsonConvert.SerializeObject(this, Formatting.Indented);
    }

}
}
